uwblueprint/planet-read
# Google OAuth Sign-In

## Notion ticket link
[Enable users to sign up with google oauth](https://www.notion.so/uwblueprintexecs/Enable-users-to-sign-up-with-google-oauth-4a23bd6c18e440ae94bae28ce6897054)


## Implementation description
Going to explain the data flow since I think that's the best way to explain it.
* Make the user sign in on the front end, from this we obtain a tokenId for that specific user.
* Send tokenId to the back end, Make a call to a Google endpoint to verify the the tokenId and get the user's info, access token, and refresh token.
* Check if the user exists in our database by email, also check if they somehow exist on the Firebase but not locally (this is likely to happen if you drop volumes or smthn)
* If the user doesn't exist, automatically make them an account using data from google
* access & refresh tokens are returned to the frontend and life is good

## Steps to test

There are 4 cases to test here.
1. Email associated with an account that was created with email & password, and the user is now logging in with that email through Google
2. Email does not have account locally or on Firebase
3. There isn't an email associated with the account locally but there is one on Firebase
4. There is an account associated with the email on both Firebase and present locally (This should be the case 99% of the time)

If you were paying attention, you might have noticed that I didn't mention to test if the record exists locally but it's not on Firebase. That's because I don't think that it should ever happen but feel free to let me know if that is a small brain idea.

All this testing should be with a **Google account** (we all have bp emails that work)

1. To test my super swaggy 100% guaranteed to work PR you'll need use the Firebase console.
    a. Go to the planetread-secrets channel on slack and get the credentials for the Google account
    b. Login to [Firebase](https://firebase.google.com/) with those credentials
    c. On the left hand side you should see Authentication, go there and you'll see all the users.
    d. For a given user if you press on the kebab menu on the right side, you'll see the option to delete that user
    e. If you're already created an account using your BP email, delete that user for the sake of testing

2. Run the container, go to the web app, and try logging in with Google.
3. Log out, and log in with Google again. The first time tested case 2, the second time tests case 4.
4. Navigate to the db using the instructions for interesting users in the README for this repo and delete the account that was just made. Login again, this tests case 3.
5. Delete that account again, now create a email & password account with the same email, login with Google, this is case 1.
6. Congratulations, you'e officially exhausted all the cases that I can think of.

## What should reviewers focus on?
* The auth service implementation and the Firebase rest client. These are where the bulk of the work is, especially the auth service. Please let me know if there's a better way to do it than I have right now.
* Themselves, take care of yourself!

![cdd](https://user-images.githubusercontent.com/53587708/125234470-d1fae600-e2ae-11eb-8b54-8a8100a35c2b.jpg)


## Checklist
